sqldelight-gradle-plugin/src/main/kotlin/com/squareup/sqldelight/gradle/SqlDelightDatabase.kt

```diff
@@ -24,25 +24,66 @@ class SqlDelightDatabase(
     get() = File(project.buildDir, "sqldelight/$name")
 
   private val sources by lazy { sources() }
+  private val dependencies = mutableListOf<SqlDelightDatabase>()
+
+  private var recursionGuard = false
 
   fun methodMissing(name: String, args: Any): Any {
     return (project as GroovyObject).invokeMethod(name, args)
   }
 
+  fun dependency(dependencyProject: Project) {
+    dependencyProject.afterEvaluate {
+      val dependency = dependencyProject.extensions.findByType(SqlDelightExtension::class.java)
+          ?: throw IllegalStateException("Cannot depend on a module with no sqldelight plugin.")
+      dependencies.add(dependency.databases.singleOrNull { it.name == name }
+          ?: throw IllegalStateException("No database named $name in $dependencyProject"))
+    }
+  }
+
   internal fun getProperties(): SqlDelightDatabaseProperties {
+    if (recursionGuard) {
+      throw IllegalStateException("Found a circular dependency in $project with database $name")
+    }
+    recursionGuard = true
+
     val packageName = requireNotNull(packageName) { "property packageName must be provided" }
```

Comment:
I'm a bit uneasy about having code between setting the value to true and the try/finally. Can we move this to the top of the function?